File: src/features/accounting/types/payroll.ts
// Payroll Types - Phase 6
// Mirrors Rust DTOs from src-tauri/src/application/dto/payroll.rs

// Payroll Run Status
export const PAYROLL_RUN_STATUS = {
  DRAFT: "draft",
  PROCESSING: "processing",
  COMPLETED: "completed",
  CANCELLED: "cancelled",
} as const;

export type PayrollRunStatus = (typeof PAYROLL_RUN_STATUS)[keyof typeof PAYROLL_RUN_STATUS];

// Payroll Entry Status
export const PAYROLL_ENTRY_STATUS = {
  PENDING: "pending",
  PROCESSED: "processed",
  PAID: "paid",
  FAILED: "failed",
} as const;

export type PayrollEntryStatus = (typeof PAYROLL_ENTRY_STATUS)[keyof typeof PAYROLL_ENTRY_STATUS];

// Run Payroll Request
export interface RunPayrollRequest {
  period_start: string;
  period_end: string;
  employee_ids: string[];
  created_by: string;
}

// Payroll Run
export interface PayrollRun {
  id: string;
  period_start: string;
  period_end: string;
  period_display: string;
  status: PayrollRunStatus;
  total_gross: number;
  total_deductions: number;
  total_net: number;
  employee_count: number;
  created_at: string;
  created_by: string;
}

// Payroll Entry
export interface PayrollEntry {
  id: string;
  payroll_run_id: string;
  employee_id: string;
  employee_name: string;
  // Income
  base_salary: number;
  hours_worked: number;
  overtime_hours: number;
  overtime_amount: number;
  bonuses: number;
  commissions: number;
  mobility: number;
  food: number;
  other_income: number;
  gross_income: number;
  // Deductions
  afp_deduction: number;
  onp_deduction: number;
  essalud: number;
  itf: number;
  other_deductions: number;
  total_deductions: number;
  // Net
  net_income: number;
  status: PayrollEntryStatus;
  created_at: string;
}

// Payroll Run with Entries
export interface PayrollRunWithEntries {
  run: PayrollRun;
  entries: PayrollEntry[];
}

// Payroll Summary
export interface PayrollSummary {
  total_payroll: number;
  total_gross: number;
  total_deductions: number;
  total_net: number;
  employee_count: number;
  latest_run?: PayrollRun;
}

// Entry filters
export interface PayrollFilters {
  period_start?: string;
  period_end?: string;
  status?: PayrollRunStatus;
  employee_id?: string;
}

File: src-tauri/src/application/dto/payroll.rs
//! Payroll DTOs

use crate::domain::entities::payroll::{PayrollEntryStatus, PayrollRunStatus};
use serde::{Deserialize, Serialize};

/// Run payroll request
#[derive(Debug, Deserialize)]
pub struct RunPayrollRequest {
    pub period_start: String,
    pub period_end: String,
    pub employee_ids: Vec<String>,
    pub created_by: String,
}

/// Payroll run response
#[derive(Debug, Serialize)]
pub struct PayrollRunDto {
    pub id: String,
    pub period_start: String,
    pub period_end: String,
    pub period_display: String,
    pub status: PayrollRunStatus,
    pub total_gross: f64,
    pub total_deductions: f64,
    pub total_net: f64,
    pub employee_count: i32,
    pub created_at: String,
    pub created_by: String,
}

impl From<crate::domain::entities::payroll::PayrollRun> for PayrollRunDto {
    fn from(run: crate::domain::entities::payroll::PayrollRun) -> Self {
        let period = run.period_display();
        Self {
            id: run.id,
            period_start: run.period_start.to_rfc3339(),
            period_end: run.period_end.to_rfc3339(),
            period_display: period,
            status: run.status,
            total_gross: run.total_gross,
            total_deductions: run.total_deductions,
            total_net: run.total_net,
            employee_count: 0, // Will be filled by service
            created_at: run.created_at.to_rfc3339(),
            created_by: run.created_by,
        }
    }
}

/// Payroll entry response
#[derive(Debug, Serialize)]
pub struct PayrollEntryDto {
    pub id: String,
    pub payroll_run_id: String,
    pub employee_id: String,
    pub employee_name: String,
    // Income
    pub base_salary: f64,
    pub hours_worked: f64,
    pub overtime_hours: f64,
    pub overtime_amount: f64,
    pub bonuses: f64,
    pub commissions: f64,
    pub mobility: f64,
    pub food: f64,
    pub other_income: f64,
    pub gross_income: f64,
    // Deductions
    pub afp_deduction: f64,
    pub onp_deduction: f64,
    pub essalud: f64,
    pub itf: f64,
    pub other_deductions: f64,
    pub total_deductions: f64,
    // Net
    pub net_income: f64,
    pub status: PayrollEntryStatus,
    pub created_at: String,
}

impl From<crate::domain::entities::payroll::PayrollEntry> for PayrollEntryDto {
    fn from(entry: crate::domain::entities::payroll::PayrollEntry) -> Self {
        Self {
            id: entry.id,
            payroll_run_id: entry.payroll_run_id,
            employee_id: entry.employee_id,
            employee_name: String::new(), // Will be filled by service
            base_salary: entry.base_salary,
            hours_worked: entry.hours_worked,
            overtime_hours: entry.overtime_hours,
            overtime_amount: entry.overtime_amount,
            bonuses: entry.bonuses,
            commissions: entry.commissions,
            mobility: entry.mobility,
            food: entry.food,
            other_income: entry.other_income,
            gross_income: entry.gross_income,
            afp_deduction: entry.afp_deduction,
            onp_deduction: entry.onp_deduction,
            essalud: entry.essalud,
            itf: entry.itf,
            other_deductions: entry.other_deductions,
            total_deductions: entry.afp_deduction
                + entry.onp_deduction
                + entry.essalud
                + entry.itf
                + entry.other_deductions,
            net_income: entry.net_income,
            status: entry.status,
            created_at: entry.created_at.to_rfc3339(),
        }
    }
}

/// Payroll run with entries
#[derive(Debug, Serialize)]
pub struct PayrollRunWithEntriesDto {
    pub run: PayrollRunDto,
    pub entries: Vec<PayrollEntryDto>,
}

/// Payroll summary for dashboard
#[derive(Debug, Serialize)]
pub struct PayrollSummary {
    pub total_payroll: f64,
    pub total_gross: f64,
    pub total_deductions: f64,
    pub total_net: f64,
    pub employee_count: i32,
    pub latest_run: Option<PayrollRunDto>,
}
